man/rvp.1.txt
RVP(1)
======
:doctype: manpage
:man source:   {manual_package}
:man version:  {manual_version}
:man manual:   {manual_title}

NAME
----
rvp - Relax NG validation pipe

SYNOPSIS
--------
*rvp* {*-q*|*-s*|*-v*|*-h*} {'schema.rnc'}

OPTIONS
-------
*-q*::
returns only error numbers, suppresses messages

*-s*::
takes less memory and runs slower

*-v*::
prints current version

*-h*::
displays usage sumary and exits

DESCRIPTION
-----------
*RVP* is abbreviation for *Relax NG Validation Pipe* It reads validation primitives from the standard input and reports result to the standard output. It's main purpose is to ease embedding of a Relax NG validator into various languages and environments. An application would launch RVP as a parallel proces and use a simple protocol to perform validation. The protocol, in BNF, is:

 query ::= (
 quit
 | start
 | start-tag-open
 | attribute
 | start-tag-close
 | text
 | end-tag) z.
 quit ::= "quit".
 start ::= "start" [gramno].
 start-tag-open ::= "start-tag-open" patno name.
 attribute ::= "attribute" patno name value.
 start-tag-close :: = "start-tag-close" patno name.
 text ::= ("text"|"mixed") patno text.
 end-tag ::= "end-tag" patno name.
 response ::= (ok | er | error) z.
 ok ::= "ok" patno.
 er ::= "er" patno erno.
 error ::= "error" patno erno error.
 z ::= "\0" .

RVP assumes that a pipe in a name separates the local part from the namespace URI (it is what one gets if specifies `|' as namespace separator to Expat).

Error codes can be grabbed from rvp sources by grep \_ER_.h and OR-ing them with corresponding masks from erbit.h. Additionally, error 0 is the protocol format error.

Either er or error responses are returned, not both; -q chooses between concise and verbose forms (invocation syntax described later).

Start passes the index of a grammar (first frammar in the list of command-line arguments has number 0); if the number is omitted, 0 is assumed.

Quit is not opposite of start; instead, it quits RVP.

EXAMPLES
--------
To assist emvessing RVP, samples in Perl and Python are provided. The scripts use Expat wrappers for each of the languages to parse documents. They take a Relax NG grammar (in the compact syntax) as the command line argument and read the XML from the standard input. For example, the following commands validate rnv.dbx against docbook.rnc:

 perl rvp.pl docbook.rnc < rnv.dbx
 python rvp.py docbook.rnc < rnv.dbx

The scripts are kept simple and unobscured to illustrate the technique, rather than being designed as general-purpose modules. Programmers using Perl, Python, Ruby and other languages are encouraged to implement and share reusable RVP-based components for languages of choice.

HOMEPAGE
--------
http://sourceforge.net/projects/rnv/

AUTHOR
------
Michael Simacek <msimacek@redhat.com>. This manpage was made from readme written by *RNV*'s author David Tolpin <dvd@davidashen.net>.

SEE ALSO
--------
rnv(1), arx(1)


ChangeLog
/* $Id$ */

1.7.11 Thu Oct  4 00:42:00 UTC+2 2012
mxp:  Fix gMonth regex pattern

1.7.10 Wed Feb  6 14:26:13 GMT 2008
hartwork: Path order fixes for configure.in
hartwork: SCM fix for configure.in
hartwork: Debian man page added

1.7.9 Tue May  1 03:12:23 GMT 2007
hartwork: GNU build files

1.7.8 Thu Nov  2 12:52:18 AMT 2006
dvd: build script for OpenVMS by Jim Duff, http://eight-cubed.com/

1.7.7 Sun Jan 29 11:13:23 PST 2006
dvd: rnv exits with non-zero return code if validation fails.

1.7.6 Fri Jan  7 01:07:37 AMT 2005
dvd: base64: I hope I got it right this time.

Thu Jan  6 15:46:55 AMT 2005
dvd: base64 regular expression checks full syntax of RFC2045.

1.7.5 Thu Jan  6 02:18:51 AMT 2005
dvd: build error fixed

1.7.4 Wed Jan  5 13:55:41 AMT 2005
dvd: unit tests for s.c and xsd.c. 

1.7.3 Wed Jan  5 05:57:55 AMT 2005
dvd: bug in xsd.c:b64cmpn fixed (thanks to Sridhar Guthula), 
     tiny code cleanup in comparison routines. 

1.7.2 Tue Nov  9 15:28:23 AMT 2004
dvd: fixed a bug in rvp.py, didn't update the current pattern after
     text or mixed

1.7.1 Thu Aug 19 00:58:58 AMST 2004
dvd: discovered and fixed a bug with parsing of builtin datatypes,
     noone seems to use them explicitely.

1.7.0 Sat Mar 27 19:26:39 AMT 2004
dvd: entities are not expanded, use XX if you need to.
     error messages have the same syntax as Jing's.

1.6.6 Sun Mar 21 22:41:10 AMT 2004
dvd: turned on parsing of external parameter entities;
     with the next next release rnv will be split in two;
     the core utility will just validate an XML stream, an optional
     preprocessor will handle entities, dtds and xinclude

1.6.5 Sat Mar 13 21:24:31 AMT 2004
dvd: continued overhaul of rn.c and other places

1.6.4 Fri Mar 12 13:15:33 AMT 2004
dvd: looked again at the code in rn.c and rnl.c, streamlined and simplified
     many things

Fri Mar 12 04:44:17 AMT 2004
dvd: better logics in mark_p

1.6.3 Thu Mar 11 22:47:17 AMT 2004
dvd: Bug fix in mark_p -- no more dangling references to unused patterns.

1.6.2 Mon Mar  8 13:22:52 AMT 2004
dvd: Added processing of davidashen-net-xg-file and -pos PIs to
     work with xg.

1.6.1 Mon Mar  1 12:55:48 AMT 2004
dvd: Added ranlib to Makefile to make rnv build smoothly on Mac OS X.

1.6.0 Thu Feb 26 21:14:39 AMT 2004
dvd: fixed mixed to be interleave, not choice. The test for this
     was wrong.

Wed Feb 25 03:59:45 AMT 2004
dvd: got rid of ##; prone to name conflicts

1.5.8 Wed Feb 18 16:53:05 AMT 2004
dvd: enhanced diagnostics: first required elements and attributes
     are reported, then optional

1.5.7 Mon Feb 16 20:03:13 AMT 2004
dvd: rewrote the pattern space compression algorithm -- it should
     be correct now, and if it is still not, debugging should be
     much easier

1.5.6 Tue Feb 10 19:37:07 AMT 2004
dvd: actually turned path checks on

1.5.5 Fri Jan 30 11:07:01 AMT 2004
dvd: added check for recursion in structured regexps

1.5.4 Fri Jan 30 00:10:32 AMT 2004
dvd: implemented structured regexps for scheme-datatypes

1.5.3 Thu Jan 29 03:19:29 AMT 2004
dvd: made it compile on SunOS 5.8 (name clashes)

1.5.2 Wed Jan 28 15:35:38 AMT 2004
dvd: added handler for external system entities
     Makefile.bcc is back; people still need win32, and cygwin is bulky.

Wed Jan 28 01:43:42 AMT 2004
dvd: removed static pointers from rnv and qualified name passed
     from Expat is modified instead. This is safe.

Mon Jan 26 21:13:30 AMT 2004
dvd: ++ moved out of assert in xsd_tm.c

1.5.1 Mon Jan 26 01:25:42 AMT 2004
dvd: Internal release, regexp in Scheme implemented

Sat Jan 24 22:42:58 AMT 2004
dvd: fixed rx.c: upper bound in quantifier

Fri Jan 23 21:51:34 AMT 2004
dvd: newRef+1 is initialized to 0 (bugfix, integrity check in ht)

Wed Jan 21 02:38:09 AMT 2004
dvd: SCM is now working as embedded engine; it was not obvious
     that the initialization call had to be lower on the stack
     than all other calls to SCM;
     static mode added to m, if M_STATIC is non-zero it is the
     size of available memory in bytes -- good for debugging.
     
Tue Jan 20 22:24:50 AMT 2004
dvd: changed rx to conform to XSD errata (- is not a valid range)
     added -d and -e to rnv; I was not sure for a long time whether
     I should do it.

Tue Jan 20 04:45:39 AMT 2004
dvd: rearranged a lot of things; put error reporting through a single
     function er_vprintf, useful for interfacing to scripting languages;
     some success with dsl -- dsl-equal? works

Tue Jan 20 00:24:56 AMT 2004
dvd: renamed strops to s, memops to m, functions has s_ and m_ prefixes

Mon Jan 19 13:43:57 AMT 2004
dvd: fixed typo in diagnostics when literal is missing after ~

Thu Jan 15 18:10:02 AMT 2004
dvd: added -d command to rvp to load external type checker;
     dxl.c dxl.h implement
     http://davidashen.net/relaxng/pluggable-datatypes datatype;
     xsdck makes xsd: pluggable.

1.4.1 Tue Jan 13 14:29:08 AMT 2004
dvd: added sys/types.h where appropriate; fixed Makefile.bcc
     to include rnl.{c,h,obj}

1.4.0 Mon Jan 12 17:39:58 AMT 2004
dvd: changed rvp.py to use os.read|os.write for performance

Mon Jan 12 03:53:56 AMT 2004
dvd: rvp is implemented; reads validation events from input, sends diagnostics
     to output; embedding examples in perl and python

Sat Jan 10 03:18:12 AMT 2004
dvd: factored out loading of rnc into rnl; now, instead of calling rnc
     and rnd separately, rnl_(fn|fd|s) is called.

1.3.5 Tue Jan  6 11:56:06 AMT 2004
dvd: learned that UTF-8 BOM is reality, added processing; fixed check for name
     characters in arx, made it >0x7f||nmtoken  (no good reason to process unicode);
     added end-of-line to error messages from Expat

1.3.4 Mon Jan  5 22:43:05 AMT 2004
dvd: grammar samples added to the distribution, entries for other grammars
     commented out in tools/arx.conf

Mon Jan  5 17:00:57 AMT 2004
dvd: isany is now ary_isany (ary is helpers for arx)

Mon Jan  5 00:56:25 AMT 2004
dvd: rnx_isany is added. It is used by arx and checks for a pattern that matches any element;

rvp.c
/* $Id$ */

/* validation pipe:
 synopsis

   rvp -qsdevh grammar.rnc

 reads from 0, writes to 1, 2 for grammar parse errors only, then redirected.
   -q switches to numerical error codes
   -s takes less space but more time
   -d plugs in an external type checker
   -e the argument is a Scheme program providing a datatype library
   -v displays version
   -h help message
 exit code: 0 on valid, non-zero on invalid

 protocol
  query ::= (start | quit | start-tag-open | attribute | start-tag-close | text | end-tag) z.
   quit ::= "quit".
   start ::= "start" [gramno].
   start-tag-open ::= "start-tag-open" patno name.
   attribute ::= "attribute" patno name value.
   start-tag-close :: = "start-tag-close" patno name.
   text ::= ("text"|"mixed") patno text.
   end-tag ::= "end-tag" patno name.
  response ::= (ok | er | error) z.
   ok ::= "ok" patno.
   er ::= "er" patno erno.
   error ::= "error" patno erno error.
  z ::= "\0" .

  conventions:
    pipe in name separates namespace uri and local part
    -q?er:error
    error==0 yields message 'protocol error' and happens when a query is not understood
    start assumes gramno=0 if the argument is omitted
*/

#include <stdlib.h>
#include <stdarg.h>
#include <stdio.h>
#include <fcntl.h>  /*open,close*/
#include <sys/types.h>
#include UNISTD_H   /*open,read,close*/
#include <string.h> /*strerror*/
#include <setjmp.h>
#include <errno.h>
#include <assert.h>
#include "m.h"
#include "s.h"
#include "erbit.h"
#include "drv.h"
#include "rnl.h"
#include "rnv.h"
#include "dxl.h"
#include "dsl.h"
#include "er.h"

extern int rn_notAllowed, drv_compact, rx_compact;

#define ATT 0
#define ENT 1
#define MIX 2
#define QUIT 3
#define START 4
#define STC 5
#define STO 6
#define TXT 7
#define NKWD 8
char *kwdtab[NKWD]={
  "attribute",
  "end-tag",
  "mixed",
  "quit",
  "start",
  "start-tag-close",
  "start-tag-open",
  "text"
};

#define OK "ok %u"
#define ER "er %u"
#define ERROR "error %u"

#define LEN_B 1024

static FILE *nstderr;
static int explain=1, lasterr, *starts, n_st;
static int len_q,n_q; char *quebuf;
static int erp[2]; /* *erp to read error messages */
static jmp_buf IOER;

static void verror_handler(int erno,va_list ap) {
  lasterr=erno;
  rnv_default_verror_handler(erno&~ERBIT_RNV,ap);
}
static void verror_handler_rnv(int erno,va_list ap) {verror_handler(erno|ERBIT_RNV,ap);}

static int initialized=0;
static void init(void) {
  if(!initialized) {initialized=1;
    rnl_init();
    rnv_init(); rnv_verror_handler=&verror_handler_rnv;
    drv_add_dtl(DXL_URL,&dxl_equal,&dxl_allows);
    drv_add_dtl(DSL_URL,&dsl_equal,&dsl_allows);
    quebuf=(char*)m_alloc(len_q=LEN_B,sizeof(char));
  }
}

static int tok(int i) {
  for(;;) {
    switch(quebuf[i]) {
    case '\t': case '\n': case '\r': case ' ': break;
    default: return i;
    }
    ++i;
  }
}

static int endtok(int i) {
  for(;;) {
    switch(quebuf[i]) {
    case '\0': case '\t': case '\n': case '\r': case ' ': return i;
    default: break;
    }
    ++i;
  }
}

static void writeall(int fd,char *buf,int len) {
  int ofs=0;
  do {
    int n=write(fd,buf+ofs,len);
    if(n==-1) longjmp(IOER,1);
    ofs+=n; len-=n;
  } while(len);
}

static void resp(int ok,int patno,int prevno) {
  int len;
  static char buf[LEN_B];
  char *f=(char*)(ok?OK:explain?ERROR:ER);
  len=sprintf(buf,f,patno); assert(len<LEN_B);
  writeall(1,buf,len);
  if(!ok) {
    len=sprintf(buf," %u",lasterr); assert(len<LEN_B);
    writeall(1,buf,len);
    if(explain) {buf[0]=' '; writeall(1,buf,1);}
  }
  for(;;) { /* read always, write if verbose */
    len=read(erp[0],buf,LEN_B);
    if(len<0) {if(errno==EAGAIN) break; else longjmp(IOER,1);}
    if(len==0) break;
    if(!ok&&explain&&prevno!=rn_notAllowed) writeall(1,buf,len);
  }
  buf[0]='\0'; writeall(1,buf,1);
}

static int query(void) {
  int i,j,n,dn, kwd, patno,prevno, ok=0;
  char *name;
  n=0;
  for(;;) {
    if(n==n_q) {
      if(len_q-n_q<LEN_B) quebuf=(char*)m_stretch(quebuf,len_q=n_q+LEN_B,n_q,sizeof(char));
      dn=read(0,quebuf+n_q,LEN_B);
      if(dn<0) longjmp(IOER,1);
      if(dn==0) {errno=EIO; longjmp(IOER,1);}
      n_q+=dn;
    }
    if(quebuf[n++]=='\0') break;
  }

  j=endtok(i=tok(0));
  if((kwd=s_ntab(quebuf+i,j-i,kwdtab,NKWD))==QUIT) {resp(1,0,0); return 0;}
  switch(kwd) {
  case START:
    j=endtok((i=tok(j)));
    patno=0; while(i!=j) patno=patno*10+quebuf[i++]-'0';
    if(patno>=n_st) goto PROTER;
    ok=1; patno=starts[patno];
    break;
  case STO: case ATT: case STC: case TXT: case MIX: case ENT:
    j=endtok((i=tok(j))); if(i==j) goto PROTER;
    patno=0; do patno=patno*10+quebuf[i++]-'0'; while(i!=j);
    if(patno==0) goto PROTER; /* 0 is ERROR, not allowed */
    switch(kwd) {
    case STO: case ATT: case STC: case ENT:
      j=endtok((i=tok(j))); if(i==j||(kwd==ATT&&quebuf[j]=='\0')) goto PROTER;
      name=quebuf+i; quebuf[j]='\0';
      switch(kwd) {
      case STO: ok=rnv_start_tag_open(&patno,&prevno,name); break;
      case ATT: ok=rnv_attribute(&patno,&prevno,name,quebuf+j+1); break;
      case STC: ok=rnv_start_tag_close(&patno,&prevno,name); break;
      case ENT: ok=rnv_end_tag(&patno,&prevno,name); break;
      }
      break;
    case TXT: case MIX:
      if(quebuf[j]) ++j; i=j; while(quebuf[j]) ++j;
      ok=rnv_text(&patno,&prevno,quebuf+i,j-i,kwd==MIX);
      break;
    }
    break;

  case NKWD: PROTER: (*er_printf)("protocol error\n"); lasterr=0; patno=0; ok=0; break;
  default: assert(0);
  }
  resp(ok,patno,prevno);

  i=0; while(n!=n_q) quebuf[i++]=quebuf[n++]; n_q=i;
  return 1;
}

static void version(void) {(*er_printf)("rvp version %s\n",RVP_VERSION);}
static void usage(void) {(*er_printf)("usage: rvp {-[qs"
#if DXL_EXC
"d"
#endif
#if DSL_SCM
"e"
#endif
"vh?]} {schema.rnc}\n");}

int main(int argc,char **argv) {
  int i, ok;
  init();

  --argc;
  while(*(++argv)&&**argv=='-') {
    --argc; i=1;
    for(;;) {
      switch(*(*argv+i)) {
      case '\0': goto END_OF_OPTIONS;
      case 'h': case '?': usage(); return 0;
      case 'v': version(); break;
      case 's': drv_compact=1; rx_compact=1; break;
#if DXL_EXC
      case 'd': dxl_cmd=*(argv+1); if(*(argv+1)) ++argv; goto END_OF_OPTIONS;
#endif
#if DSL_SCM
      case 'e': dsl_ld(*(argv+1)); if(*(argv+1)) ++argv; goto END_OF_OPTIONS;
#endif
      case 'q': explain=0; break;
      default: (*er_printf)("unknown option '-%c'\n",*(*argv+i)); break;
      }
      ++i;
    }
    END_OF_OPTIONS:;
  }

  if(*argv==NULL) {usage(); return 1;}

  starts=(int*)m_alloc(argc,sizeof(int));
  ok=1; n_st=0;
  do {
    ok=(starts[n_st++]=rnl_fn(*(argv++)))&&ok;
  } while(*argv);
  if(ok) {
    int fd2;

    nstderr=stderr;
    if(setjmp(IOER)) {
      fprintf(nstderr,"%s\n",strerror(errno));
      return EXIT_FAILURE;
    }

    if((fd2=dup(2))==-1) longjmp(IOER,1);
    nstderr=fdopen(fd2,"w");
    if(pipe(erp)==-1||dup2(erp[1],2)==-1) longjmp(IOER,1);
    fcntl(erp[0],F_SETFL,O_NONBLOCK);
    setbuf(stderr,NULL);

    while(query());
    return EXIT_SUCCESS;
  }
  return EXIT_FAILURE;
}
